### PR TITLE
Split Sidecar to it's own container

### PR DESCRIPTION
**Pull request by dvaldivia**
Right now we have a problem where upgrading/downgrading the operator causes all tenants to perform a rolling restart, this is because the sidecar image is the same as the operator.

This PR splits the sidecar into it's own binary, intended to become it's own container image with separate release cycle.

This way, only when there's changes to the sidecar code, we issue a new release and update operator so we can warn users that any given operator release will cause a rolling restart of all tenants.

### CODE REVIEW

File: pkg/resources/statefulsets/minio-sidecar.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// This file is part of MinIO Operator
+// Copyright (c) 2023 MinIO, Inc.
```
Comment (shtripat) [inline pkg/resources/statefulsets/minio-sidecar.go:2]:
```suggestion
// Copyright (c) 2024 MinIO, Inc.
```

File: sidecar/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      release="${TAG}" \
+      summary="MinIO Operator brings native support for MinIO, Console, and Encryption to Kubernetes." \
+      description="MinIO object storage is fundamentally different. Designed for performance and the S3 API, it is 100% open-source. MinIO is ideal for large, private cloud environments with stringent security requirements and delivers mission-critical availability across a diverse range of workloads."
+
+# On RHEL the certificate bundle is located at:
+# - /etc/pki/tls/certs/ca-bundle.crt (RHEL 6)
+# - /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem (RHEL 7)
+COPY --from=build /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem /etc/pki/ca-trust/extracted/pem/
+
+#COPY CREDITS /licenses/CREDITS
```
Comment (shtripat) [inline sidecar/Dockerfile:22]: needed?

File: sidecar/cmd/sidecar/app_commands.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// This file is part of MinIO Operator
+// Copyright (c) 2023 MinIO, Inc.
```
Comment (shtripat) [inline sidecar/cmd/sidecar/app_commands.go:2]:
```suggestion
// Copyright (c) 2024 MinIO, Inc.
```

File: sidecar/cmd/sidecar/main.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
+// This file is part of MinIO Operator
+// Copyright (c) 2023 MinIO, Inc.
```
Comment (shtripat) [inline sidecar/cmd/sidecar/main.go:2]:
```suggestion
// Copyright (c) 2024 MinIO, Inc.
```

File: sidecar/cmd/sidecar/main.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ {{end}}{{end}}
+VERSION:
+ {{.Version}}
+`
+
+func newApp(name string) *cli.App {
+	// Collection of console commands currently supported are.
+	var commands []cli.Command
+
+	// Collection of console commands currently supported in a trie tree.
```
Comment (shtripat) [inline sidecar/cmd/sidecar/main.go:56]:
```suggestion
	// Collection of console commands currently supported in a tree.
```

File: sidecar/cmd/sidecar/main.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+		Usage: "show help",
+	}
+
+	app := cli.NewApp()
+	app.Name = name
+	app.Version = pkg.Version + " - " + pkg.ShortCommitID
+	app.Author = "MinIO, Inc."
+	app.Usage = "MinIO Operator Sidecar"
+	app.Description = `MinIO Operator automates the orchestration of MinIO Tenants on Kubernetes.`
+	app.Copyright = "(c) 2023 MinIO, Inc."
```
Comment (shtripat) [inline sidecar/cmd/sidecar/main.go:102]:
```suggestion
	app.Copyright = "(c) 2024 MinIO, Inc."
```

File: sidecar/cmd/sidecar/sidecar.go [hunk 7]
```diff
@@ -1,16 +1,18 @@
-// Copyright (C) 2023, MinIO, Inc.
+// This file is part of MinIO Operator
+// Copyright (c) 2023 MinIO, Inc.
```
Comment (shtripat) [inline sidecar/cmd/sidecar/sidecar.go:2]:
```suggestion
// Copyright (c) 2024 MinIO, Inc.
```

File: sidecar/pkg/build-constants.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
+// This file is part of MinIO Operator
+// Copyright (c) 2023 MinIO, Inc.
```
Comment (shtripat) [inline sidecar/pkg/build-constants.go:2]:
```suggestion
// Copyright (c) 2024 MinIO, Inc.
```

File: sidecar/pkg/sidecar/webhook_server.go [hunk 9]
```diff
@@ -1,25 +1,26 @@
-// Copyright (C) 2020, MinIO, Inc.
+// This file is part of MinIO Operator
+// Copyright (c) 2023 MinIO, Inc.
```
Comment (shtripat) [inline sidecar/pkg/sidecar/webhook_server.go:2]:
```suggestion
// Copyright (c) 2024 MinIO, Inc.
```

File: sidecar/pkg/validator/validator.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
+// This file is part of MinIO Operator
+// Copyright (c) 2023 MinIO, Inc.
```
Comment (shtripat) [inline sidecar/pkg/validator/validator.go:2]:
```suggestion
// Copyright (c) 2024 MinIO, Inc.
```

File: sidecar/pkg/validator/validator.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+		panic(err)
+	}
+
+	namespace := miniov2.GetNSFromFile()
+
+	cfg, err := rest.InClusterConfig()
+	// If config is passed as a flag use that instead
+	//if kubeconfig != "" {
+	//	cfg, err = clientcmd.BuildConfigFromFlags(masterURL, kubeconfig)
+	//}
```
Comment (shtripat) [inline sidecar/pkg/validator/validator.go:49]: needed?

File: testing/common.sh [hunk 12]
```diff
@@ -557,6 +557,14 @@ function install_operator() {
   echo 'start - load compiled image so we can use it later on'
   try kind load docker-image minio/operator:noop
   echo 'end - load compiled image so we can use it later on'
+  # To compile current branch
+  echo "Compiling Current Branch Sidecar"
+  TAG=minio/operator-sidecar:noop
+  (cd "${SCRIPT_DIR}/../sidecar" && try docker build -t $TAG .) # will not change your shell's current directory
+
+  echo 'start - load compiled sidecar  image so we can use it later on'
```
Comment (shtripat) [inline testing/common.sh:565]:
```suggestion
  echo 'start - load compiled sidecar image so we can use it later on'
```

File: sidecar/go.mod [hunk 13]
```diff
@@ -0,0 +1,174 @@
+module github.com/minio/operator/sidecar
+
+go 1.22.2
```
Comment (ramondeklein) [inline sidecar/go.mod:3]: In `.github/workflows/kubernetes-tests.yml` it uses Go 1.21.

File: sidecar/pkg/validator/validator.go (hunk 11 above)
Comment (ramondeklein) [reply sidecar/pkg/validator/validator.go:49]: I guess this is useful to uncomment while debugging.

**Review by ramondeklein — APPROVED** [review]
Do I understand correctly that the sidecar still uses code from the operator (it references `github.com/minio/operator` in `go.mod`)? It looks like, although that code is in the same repository, it will use a fixed version that is fetched from GitHub? So, if changes are made to `github.com/minio/operator` that should be used in the sidecar, then these changes need to be pushed to GitHub and the sidecar dependency should be updated. Am I right?

**Comment by harshavardhana** [comment]
> This way, only when there's changes to the sidecar code, we issue a new release and update operator so we can warn users that any given operator release will cause a rolling restart of all tenants.

Is there no way to avoid this?

**Comment by harshavardhana** [comment]
what if we move the sidecar process outside of operator instead move it into a separate project?

**Comment by dvaldivia** [comment]
@harshavardhana the symbiosis between operator and sidecar is too strong, I anticipate only when we need changes to operator that we will pair them with sidecar changes in that way we can in a single PR perform both

File: .github/workflows/kubernetes-tests.yml [hunk 14]
```diff
@@ -127,11 +127,35 @@ jobs:
       - name: Operator Binary
         run: |
           make operator
+  sidecar:
+    timeout-minutes: 30
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        go-version: [ 1.21.x ]
+        os: [ ubuntu-latest ]
```
Comment (pjuarezd) [inline .github/workflows/kubernetes-tests.yml:136]: Intended to add a test here?

File: sidecar/Dockerfile (hunk 2 above)
Comment (pjuarezd) [reply sidecar/Dockerfile]: yeap, added now

**Review by pjuarezd — DISMISSED** [review]

**Review by jiuker — DISMISSED** [review]

**Review by shtripat — DISMISSED** [review]
LGTM. One open question from @ramondeklein

File: sidecar/go.mod (hunk 13 above)
Comment (harshavardhana) [reply sidecar/go.mod:3]: Let's keep this `go 1.21`

Comment (harshavardhana) [inline sidecar/go.mod:3]:
```suggestion
go 1.21
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by pjuarezd — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
BTW lets move CI/CD to go1.22.x // @allanrogerr can you send a fix?

**Pull request merged**